### PR TITLE
[Will not merge] Show the IR of the partial-auto boundary.

### PR DESCRIPTION
**Pull request by yliu120**
Command to run:

bazel test -c opt --test_filter="*test_partial_auto_ppermute" --test_output=all //tests:shard_map_test_gpu

### CODE REVIEW

**Comment by yliu120** [comment]
@pschuh @hawkinsp @yashk2810 @mattjj Please run this code and take a look. Thanks

**Comment by yliu120** [comment]
```
def test_partial_auto_ppermute(self):
    if config.use_shardy_partitioner.value:
      self.skipTest('Shardy does not support full-to-shard.')

    mesh = jtu.create_mesh((4, 2), ('i', 'j'))
    x = jnp.arange(8.).reshape(4, 2)

    def g(x):
      x = jax.lax.with_sharding_constraint(x, NamedSharding(mesh, P(None, 'j')))
      return jax.lax.ppermute(x, 'i', [(0, 1), (1, 2), (2, 3), (3, 0)])

    @jax.jit
    def f(x):
      return shard_map(g,
                       mesh, in_specs=P('i'), out_specs=P('i'),
                       check_rep=False, auto=frozenset({'j'}))(x)

    s = NamedSharding(mesh, P("i", "j"))
    compiled = jax.jit(f, in_shardings=s, out_shardings=s).lower(x).compile()  # don't crash
    print(compiled.as_text())
```

This example works.

**Comment by yliu120** [comment]
But these examples are a bit inefficient by adding an all-gather and dynamic-slice,

```
def test_partial_auto_ppermute(self):
    if config.use_shardy_partitioner.value:
      self.skipTest('Shardy does not support full-to-shard.')

    mesh = jtu.create_mesh((4, 2), ('i', 'j'))
    x = jnp.arange(8.)

    def g(x):
      x = jax.lax.with_sharding_constraint(x, NamedSharding(mesh, P('j')))
      return jax.lax.ppermute(x, 'i', [(0, 1), (1, 2), (2, 3), (3, 0)])

    @jax.jit
    def f(x):
      return shard_map(g,
                       mesh, in_specs=P('i'), out_specs=P('i'),
                       check_rep=False, auto=frozenset({'j'}))(x)

    s = NamedSharding(mesh, P(("i", "j")))
    compiled = jax.jit(f, in_shardings=s, out_shardings=s).lower(x).compile()  # don't crash
    print(compiled.as_text())
```

```
def test_partial_auto_ppermute(self):
    if config.use_shardy_partitioner.value:
      self.skipTest('Shardy does not support full-to-shard.')

    mesh = jtu.create_mesh((4, 2), ('i', 'j'))
    x = jnp.arange(8.).reshape(4, 2)

    def g(x):
      x = x.reshape(-1)
      x = jax.lax.with_sharding_constraint(x, NamedSharding(mesh, P('j')))
      return jax.lax.ppermute(x, 'i', [(0, 1), (1, 2), (2, 3), (3, 0)])

    @jax.jit
    def f(x):
      permuted = shard_map(g,
          mesh, in_specs=P('i'), out_specs=P('i'),
          check_rep=False, auto=frozenset({'j'}))(x)
      return permuted.reshape(4, 2)

    s = NamedSharding(mesh, P("i", "j"))
    compiled = jax.jit(f, in_shardings=s, out_shardings=s).lower(x).compile()  # don't crash
    print(compiled.as_text())
```

**Comment by bixia1** [comment]
This turns out to be a problem in GSPMD but not Shardy, the new partitioner. If you change [this flag](https://github.com/jax-ml/jax/blob/10bb38bb79aae052d56ec6074cd74eda1fc40531/jax/_src/config.py#L1685-L1687) to true, such as just flipping the default value to true, the optimized HLO generated by the GPU compiler is like this

```
ENTRY main.12_spmd {
  param.1 = f64[1]{0} parameter(0), sharding={devices=[8]<=[8]}, metadata={op_name="x"}
  collective-permute-start = (f64[1]{0}, f64[1]{0}) collective-permute-start(param.1), channel_id=1, source_target_pairs={{0,2},{2,4},{4,6},{6,0},{1,3},{3,5},{5,7},{7,1}}, metadata={op_name="jit(f)/jit(main)/ppermute" source_file="third_party/py/jax/tests/pjit_test.py" source_line=1332}, backend_config={"operation_queue_id":"0","wait_on_operation_queues":[],"collective_backend_config":{"is_sync":true,"no_parallel_custom_call":false,"is_pipelined":false},"force_earliest_schedule":false}
  ROOT collective-permute-done = f64[1]{0} collective-permute-done(collective-permute-start), metadata={op_name="jit(f)/jit(main)/ppermute" source_file="third_party/py/jax/tests/pjit_test.py" source_line=1332}
}
```

**Comment by Vjudith77** [comment]
Judith Valenzuela

**Comment by yliu120** [comment]
> self.skipTest('Shardy does not support full-to-shard.')

Does Shardy support FullToShard now?

**Comment by yliu120** [comment]
My point is that we still want to fix the old GSPMD impl for this thing.
shardy is so new and it is very risky to use unless we can make sure it is compatible with all JAX existing features.

**Comment by yliu120** [comment]
I just reported to Parker @pschuh that axis_index not working for shardy inside partial auto.

**Comment by bixia1** [comment]
> > self.skipTest('Shardy does not support full-to-shard.')
> 
> Does Shardy support FullToShard now?

Shardy definitely support [these FullToShard ShardToFull ops](https://github.com/jax-ml/jax/blob/fc9356085edca50cae8b969ad1108a2ab3c2db71/jax/_src/interpreters/mlir.py#L2591-L2592) .

There may be something missing in the path, let me find out what it will take to enable these tests.

**Comment by bartchr808** [comment]
Hey! So Shardy was doing the "right" thing only because of a bug unfortunately. This was fixed in

https://github.com/jax-ml/jax/commit/4997e45743e3b243ef153674a11a826843ab37b0

The issue is that I incorrectly implemented Shardy into JAX, and what was happening was we allowed the out_sharding `P('i')` to become `P(('i', 'j'))` - which matches your `with_sharding_constraint` in the repo.

JAX doesn't allow this behavior at the moment, which is why in the PR above I didn't allow any axes to be propagated to the out-sharding - now Shardy and GSPMD have the bad all-gather and dynamic-slice.

The reason why you are seeing this is because axis `j` propagated inside, to the output, but you have a fixed `P('i')` out sharding. So you need the all-gather. Now the `with_sharding_constraint` asks for it to be further sharded on `j` which is why you get the slice.

So instead, could you update your use of shard_map to specify an `out_shardings=P(('i', 'j'))` instead? Or do you actually want the `j` to optionally propagate through the out_sharding?

**Comment by yliu120** [comment]
> Hey! So Shardy was doing the "right" thing only because of a bug unfortunately. This was fixed in
> 
> [4997e45](https://github.com/jax-ml/jax/commit/4997e45743e3b243ef153674a11a826843ab37b0)
> 
> The issue is that I incorrectly implemented Shardy into JAX, and what was happening was we allowed the out_sharding `P('i')` to become `P(('i', 'j'))` - which matches your `with_sharding_constraint` in the repo.
> 
> JAX doesn't allow this behavior at the moment, which is why in the PR above I didn't allow any axes to be propagated to the out-sharding - now Shardy and GSPMD have the bad all-gather and dynamic-slice.
> 
> The reason why you are seeing this is because axis `j` propagated inside, to the output, but you have a fixed `P('i')` out sharding. So you need the all-gather. Now the `with_sharding_constraint` asks for it to be further sharded on `j` which is why you get the slice.
> 
> So instead, could you update your use of shard_map to specify an `out_shardings=P(('i', 'j'))` instead? Or do you actually want the `j` to optionally propagate through the out_sharding?

Actually this is not quite aligned with the intention of shard_map partial auto. You can consider the shard_map partial auto takes some function that passed by the users:

```
def transform(f, *args):
   def _transformed(*transformed_args):
     axis_index = jax.lax.axis_index("i")
     # Process the t_args
     return axis_index + f(*transformed_args)

   new_args = reshape(*args) # add one more dimension to the left.
   g = shard_map(_transformed, input_pspecs=P("i"), output_pspecs=P("i"), auto=frozenset({...}))
   return g
```

In this example, the `f` is a function provide by the users, and its inputs are all shareded with GSPMD. We have *no way* to specify the shardings for all its inputs.

I am wondering if we can see how everything gets sharded, then having shard_map partial auto will not be meaningful.

**Comment by ZixuanJiang** [comment]
After [this commit](https://github.com/jax-ml/jax/commit/d7c22eb1f1aaf4cf41e2e0aa103d8c67ab206daa), with Shardy, we can avoid the redundant all-gather + dynamic-slice instructions since we can update the in/out shardings for manual computation. 

With GSPMD, we still have the redundant operations. However, we will not pay attention to it since we will deprecate it soon.

Thank you!
